[WebXR+WebGPU] Bringup red triangle

mwyrzykowski opened the pull request:
#### 96cbf755f1848da5261f32815e3c99b39557344e
<pre>
[WebXR+WebGPU] Bringup red triangle
<a href="https://bugs.webkit.org/show_bug.cgi?id=278171">https://bugs.webkit.org/show_bug.cgi?id=278171</a>
<a href="https://rdar.apple.com/133958972">rdar://133958972</a>

Reviewed by Dan Glastonbury.

Move WebXR WebGPU IPC calls behind the feature flag and some changes to
render to the MTLTexture from CompositorServices.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/Modules/WebGPU/GPUTextureViewDescriptor.idl:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUConvertToBackingContext.cpp:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUConvertToBackingContext.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRBindingImpl.cpp:
(WebCore::WebGPU::XRBindingImpl::getViewSubImage):
(WebCore::WebGPU::XRBindingImpl::getPreferredColorFormat):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRBindingImpl.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRProjectionLayerImpl.cpp:
(WebCore::WebGPU::XRProjectionLayerImpl::startFrame):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRProjectionLayerImpl.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRSubImageImpl.cpp:
(WebCore::WebGPU::XRSubImageImpl::colorTexture):
(WebCore::WebGPU::XRSubImageImpl::depthStencilTexture):
(WebCore::WebGPU::XRSubImageImpl::motionVectorTexture):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRSubImageImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRBinding.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRProjectionLayer.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRSubImage.h:
* Source/WebCore/Modules/webxr/WebXRLayer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
* Source/WebCore/Modules/webxr/WebXRRenderState.cpp:
(WebCore::WebXRRenderState::setLayers):
* Source/WebCore/Modules/webxr/WebXRRenderState.h:
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::updateRenderState):
(WebCore::WebXRSession::frameShouldBeRendered const):
(WebCore::WebXRSession::onFrame):
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
(WebCore::WebXRWebGLLayer::startFrame):
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.h:
* Source/WebCore/Modules/webxr/XRGPUBinding.cpp:
(WebCore::XRGPUBinding::XRGPUBinding):
(WebCore::XRGPUBinding::device):
(WebCore::XRGPUBinding::getViewSubImage):
(WebCore::XRGPUBinding::getPreferredColorFormat):
* Source/WebCore/Modules/webxr/XRGPUBinding.h:
* Source/WebCore/Modules/webxr/XRGPUProjectionLayerInit.h:
* Source/WebCore/Modules/webxr/XRGPUSubImage.cpp:
(WebCore::XRGPUSubImage::XRGPUSubImage):
(WebCore::m_viewport):
(WebCore::textureDescriptor):
(WebCore::XRGPUSubImage::colorTexture):
(WebCore::XRGPUSubImage::depthStencilTexture):
(WebCore::XRGPUSubImage::motionVectorTexture):
(WebCore::XRGPUSubImage::getViewDescriptor const):
(WebCore::XRGPUSubImage::viewport const):
* Source/WebCore/Modules/webxr/XRGPUSubImage.h:
(WebCore::XRGPUSubImage::create):
(WebCore::XRGPUSubImage::colorTexture const): Deleted.
(WebCore::XRGPUSubImage::depthStencilTexture const): Deleted.
(WebCore::XRGPUSubImage::motionVectorTexture const): Deleted.
* Source/WebCore/Modules/webxr/XRGPUSubImage.idl:
* Source/WebCore/Modules/webxr/XRProjectionLayer.cpp:
(WebCore::XRProjectionLayer::startFrame):
* Source/WebCore/Modules/webxr/XRProjectionLayer.h:
* Source/WebCore/Modules/webxr/XRSubImage.h:
(WebCore::XRSubImage::viewport const): Deleted.
* Source/WebCore/Modules/webxr/XRWebGLSubImage.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::Device::getXRViewSubImage):
(WebGPU::Device::generateAValidationError):
* Source/WebGPU/WebGPU/MetalSPI.h:
* Source/WebGPU/WebGPU/Queue.h:
(WebGPU::Queue::setCommittedSignalEvent):

Canonical link: <a href="https://commits.webkit.org/283478@main">https://commits.webkit.org/283478@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c65d4b70efd41113494937fbd5175c11719c048b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66379 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19000 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70412 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16990 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53553 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17272 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11854 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69446 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42179 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57460 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38850 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14848 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15865 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/59493 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60744 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15190 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72115 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/65624 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10336 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14566 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60560 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10368 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57529 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60874 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14636 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8533 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2142 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/87390 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41561 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15364 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42638 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43821 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42381 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/Modules/webxr/XRProjectionLayer.cpp:
```diff
@@ -40,9 +42,18 @@ XRProjectionLayer::XRProjectionLayer(ScriptExecutionContext& scriptExecutionCont
+void XRProjectionLayer::startFrame(PlatformXR::FrameData& data)
 {
-    m_backing->startFrame();
+    auto it = data.layers.find(1); // m_framebuffer->handle());
```
Not sure if this is correct, but I used 1 based on:
```
frameData.layers.set(defaultLayerHandle(), WTFMove(layerDataRef));
```

mwyrzykowski replied:
from PlatformXRCompositor.cpp

achan00 replied:
I wonder how we can avoid hardcoding this. Right now with `WebXRWebGLLayer`, it'll ask the `PlatformXR::Device` to create the frame buffer for the layer which returns a handle (which on visionOS, it'd just return `PlatformXRCoordinator::defaultLayerHandle()`, which is `1`). Maybe the `XRProjectionLayer` can also set up its layer handle during its initialization?

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -262,6 +262,9 @@
     [m_createdNotCommittedBuffers removeObject:commandBuffer];
     ++m_submittedCommandBufferCount;
+#if !PLATFORM(IOS_FAMILY_SIMULATOR)
+    [m_commitSharedEvent setSignaledValue:m_submittedCommandBufferCount];
```
should probably be placing this in the command buffer via `encodeSignalEvent:`

mwyrzykowski replied:
`m_submittedCommandBufferCount` is incorrect in the general case, because CompositorServices expects us to count from zero

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -258,12 +258,23 @@
+        [commandBuffer encodeSignalEvent:m_commitSharedEvent value:m_commitFrameIndex];
+        m_commitSharedEvent = nil;
+    }
+//        m_commitSharedEvent.signaledValue = frameIndex;
```
should remove this

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -258,12 +258,23 @@
         });
     }];
 
+    if (m_commitSharedEvent) {
```
This works for the test page but is incorrect in general. The correct thing I think to do is:

* associate the shared event with the render target
* encode the signal event when both textures in the array have all their commands encoded.

That might still be incorrect however, since I don't think there is anything preventing a website from rendering to the same texture multiple times in a single frame.

achan00 commented on Source/WebGPU/WebGPU/XRProjectionLayer.h:
```diff
@@ -57,12 +57,22 @@ class XRProjectionLayer : public WGPUXRProjectionLayerImpl, public RefCounted<XR
+    NSMutableDictionary<NSNumber*, id<MTLTexture>>* m_depthTextures { nil };
+    id<MTLTexture> m_colorTexture { nil };
+    id<MTLTexture> m_depthTexture { nil };
+    id<MTLSharedEvent> m_sharedEvent { nil };
```
Do these need to be `RetainPtr`s?

mwyrzykowski replied:
No because WebGPU uses ARC

achan00 commented on Source/WebGPU/WebGPU/XRProjectionLayer.mm:
```diff
@@ -51,7 +54,55 @@
+    id<MTLDevice> device = m_device->device();
+    NSNumber* textureKey = @(reusableTextureIndex);
+    if (colorBuffer.sendRight()) {
+        MTLSharedTextureHandle* m_colorHandle = [[MTLSharedTextureHandle alloc] initWithMachPort:colorBuffer.sendRight()];
```
Should this be `auto colorHandle = adoptNS([[MTLSharedTextureHandle alloc] initWithMachPort:colorBuffer.sendRight()]);`?

mwyrzykowski replied:
WebGPU uses ARC so it would be fine but is not needed

achan00 commented on Source/WebGPU/WebGPU/XRProjectionLayer.mm:
```diff
@@ -51,7 +54,55 @@
+        m_colorTexture = [m_colorTextures objectForKey:textureKey];
+
+    if (depthBuffer.sendRight()) {
+        MTLSharedTextureHandle* m_depthHandle = [[MTLSharedTextureHandle alloc] initWithMachPort:depthBuffer.sendRight()];
```
Ditto.

achan00 commented on Source/WebGPU/WebGPU/XRProjectionLayer.mm:
```diff
@@ -51,7 +54,55 @@
+    NSNumber* textureKey = @(reusableTextureIndex);
+    if (colorBuffer.sendRight()) {
+        MTLSharedTextureHandle* m_colorHandle = [[MTLSharedTextureHandle alloc] initWithMachPort:colorBuffer.sendRight()];
+        m_colorTexture = [device newSharedTextureWithHandle:m_colorHandle];
```
If `m_colorTexture` is a `RetainPtr`, should this use `adoptNS(...)`?

mwyrzykowski commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2805,7 +2805,7 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, bool supportsNonP
     auto endArrayLayer = checkedSum<uint32_t>(descriptor.baseArrayLayer, descriptor.arrayLayerCount);
     if (endArrayLayer.hasOverflowed() || endArrayLayer.value() > arrayLayerCount())
-        return ERROR_STRING(@"endArrayLayer is not valid");
+        return ERROR_STRING([NSString stringWithFormat:@"endArrayLayer(%u) is not valid. Base texture array count is %u", endArrayLayer.value(), arrayLayerCount()]);
```
This was added because I ran into this error bringing up the red triangle and the error message was not sufficient descriptive to understand what was wrong without debugging

djg commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRBindingImpl.cpp:
```diff
@@ -60,15 +60,15 @@ RefPtr<XRSubImage> XRBindingImpl::getSubImage(XRProjectionLayer&, WebCore::WebXR
 TextureFormat XRBindingImpl::getPreferredColorFormat()
 {
-    return TextureFormat::Bgra8unorm;
+    return TextureFormat::Bgra8unormSRGB;
```
This might be an issue. In WebGL-based WebXR we did this and the WebGL spec assumes that it's Bgra8unorm by default so the rendering was coming out lighter than expected.

mwyrzykowski replied:
In theory this would be a website bug as the website should use the color format from getPreferredColorFormat() according to https://github.com/immersive-web/WebXR-WebGPU-Binding/blob/main/explainer.md#webgpu-binding

Does the WebXR WebGL variant have a similar preferred texture format function? I couldn't find it in https://www.w3.org/TR/webxr/ but I am not so familiar

djg commented on Source/WebCore/Modules/webxr/XRGPUSubImage.cpp:
```diff
@@ -28,13 +28,84 @@
+
 namespace WebCore {
 
+static constexpr auto gpuSubImageWidth = 1920;
+static constexpr auto gpuSubImageHeight = 1824;
```
Magic numbers!

mwyrzykowski replied:
oh yes, I should get these from CompositorServices instead 👍

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebXR+WebGPU] Bringup red triangle
https://bugs.webkit.org/show_bug.cgi?id=278171
rdar://133958972

Reviewed by Dan Glastonbury.

Move WebXR WebGPU IPC calls behind the feature flag and some changes to
render to the MTLTexture from CompositorServices.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/Modules/WebGPU/GPUTextureViewDescriptor.idl:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUConvertToBackingContext.cpp:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUConvertToBackingContext.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRBindingImpl.cpp:
(WebCore::WebGPU::XRBindingImpl::getViewSubImage):
(WebCore::WebGPU::XRBindingImpl::getPreferredColorFormat):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRBindingImpl.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRProjectionLayerImpl.cpp:
(WebCore::WebGPU::XRProjectionLayerImpl::startFrame):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRProjectionLayerImpl.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRSubImageImpl.cpp:
(WebCore::WebGPU::XRSubImageImpl::colorTexture):
(WebCore::WebGPU::XRSubImageImpl::depthStencilTexture):
(WebCore::WebGPU::XRSubImageImpl::motionVectorTexture):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRSubImageImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRBinding.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRProjectionLayer.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRSubImage.h:
* Source/WebCore/Modules/webxr/WebXRLayer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
* Source/WebCore/Modules/webxr/WebXRRenderState.cpp:
(WebCore::WebXRRenderState::setLayers):
* Source/WebCore/Modules/webxr/WebXRRenderState.h:
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::updateRenderState):
(WebCore::WebXRSession::frameShouldBeRendered const):
(WebCore::WebXRSession::onFrame):
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
(WebCore::WebXRWebGLLayer::startFrame):
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.h:
* Source/WebCore/Modules/webxr/XRGPUBinding.cpp:
(WebCore::XRGPUBinding::XRGPUBinding):
(WebCore::XRGPUBinding::device):
(WebCore::XRGPUBinding::getViewSubImage):
(WebCore::XRGPUBinding::getPreferredColorFormat):
* Source/WebCore/Modules/webxr/XRGPUBinding.h:
* Source/WebCore/Modules/webxr/XRGPUProjectionLayerInit.h:
* Source/WebCore/Modules/webxr/XRGPUSubImage.cpp:
(WebCore::XRGPUSubImage::XRGPUSubImage):
(WebCore::m_viewport):
(WebCore::textureDescriptor):
(WebCore::XRGPUSubImage::colorTexture):
(WebCore::XRGPUSubImage::depthStencilTexture):
(WebCore::XRGPUSubImage::motionVectorTexture):
(WebCore::XRGPUSubImage::getViewDescriptor const):
(WebCore::XRGPUSubImage::viewport const):
* Source/WebCore/Modules/webxr/XRGPUSubImage.h:
(WebCore::XRGPUSubImage::create):
(WebCore::XRGPUSubImage::colorTexture const): Deleted.
(WebCore::XRGPUSubImage::depthStencilTexture const): Deleted.
(WebCore::XRGPUSubImage::motionVectorTexture const): Deleted.
* Source/WebCore/Modules/webxr/XRGPUSubImage.idl:
* Source/WebCore/Modules/webxr/XRProjectionLayer.cpp:
(WebCore::XRProjectionLayer::startFrame):
* Source/WebCore/Modules/webxr/XRProjectionLayer.h:
* Source/WebCore/Modules/webxr/XRSubImage.h:
(WebCore::XRSubImage::viewport const): Deleted.
* Source/WebCore/Modules/webxr/XRWebGLSubImage.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::Device::getXRViewSubImage):
(WebGPU::Device::generateAValidationError):
* Source/WebGPU/WebGPU/MetalSPI.h:
* Source/WebGPU/WebGPU/Queue.h:
(WebGPU::Queue::setCommittedSignalEvent):

Canonical link: https://commits.webkit.org/283478@main
